File: src/app_legacy/toon.py
# app/toon.py
import re
from typing import List, Dict, Any, Tuple


ToonRow = Dict[str, Any]
ToonSchema = List[str]

DEFAULT_SEP = ";"


def _escape_field(raw: Any, sep: str = DEFAULT_SEP) -> str:
    """Escape a single field for TOON."""
    s = "" if raw is None else str(raw)
    needs_quote = (
        sep in s
        or "\n" in s
        or "\r" in s
        or '"' in s
        or s.startswith("#")
    )
    if not needs_quote:
        return s
    return '"' + s.replace('"', '""') + '"'


def _unescape_field(s: str) -> str:
    """Unescape a single TOON field back to plain string."""
    if len(s) >= 2 and s.startswith('"') and s.endswith('"'):
        inner = s[1:-1]
        return inner.replace('""', '"')
    return s


def encode_toon(rows: List[ToonRow], schema: ToonSchema, sep: str = DEFAULT_SEP) -> str:
    """
    JSON[] -> TOON string

    Header format:
        #TOON v=1 sep=; cols=item,qty,length,width,room
    """
    header = f"#TOON v=1 sep={sep} cols={','.join(schema)}"
    body_lines: List[str] = []
    for row in rows:
        parts = [_escape_field(row.get(k), sep) for k in schema]
        body_lines.append(sep.join(parts))
    return header + "\n" + "\n".join(body_lines)


def _parse_header(header: str) -> Tuple[ToonSchema, str]:
    """
    Parse header line into (schema, sep).

    Example header:
        #TOON v=1 sep=; cols=item,qty,length,width,room
    """
    cols_match = re.search(r"cols=([^\s]+)", header)
    sep_match = re.search(r"sep=([^\s]+)", header)
    if not cols_match or not sep_match:
        raise ValueError("Malformed TOON header")
    schema_str = cols_match.group(1)
    sep = sep_match.group(1)
    schema = schema_str.split(",")
    return schema, sep


def _split_line(line: str, sep: str) -> List[str]:
    """
    Split a line into fields, respecting quotes and escaped quotes.
    Assumes 1-character separator (e.g. ';').
    """
    fields: List[str] = []
    cur = []
    in_quotes = False
    i = 0
    while i < len(line):
        c = line[i]
        if in_quotes:
            if c == '"':
                # possible escaped quote
                if i + 1 < len(line) and line[i + 1] == '"':
                    cur.append('"')
                    i += 1
                else:
                    in_quotes = False
            else:
                cur.append(c)
            i += 1
            continue

        if c == '"':
            in_quotes = True
            i += 1
            continue

        if len(sep) == 1 and c == sep:
            fields.append("".join(cur))
            cur = []
        else:
            cur.append(c)
        i += 1

    fields.append("".join(cur))
    return [_unescape_field(f) for f in fields]


def decode_toon(toon: str) -> List[ToonRow]:
    """
    TOON string -> JSON[]

    Raises ValueError if header is missing/invalid.
    """
    if not toon:
        raise ValueError("Empty TOON string")

    lines = toon.replace("\r", "").split("\n")
    if not lines or not lines[0].startswith("#TOON"):
        raise ValueError("Invalid TOON header")

    header = lines[0]
    schema, sep = _parse_header(header)

    rows: List[ToonRow] = []
    for line in lines[1:]:
        line = line.strip()
        if not line:
            continue
        parts = _split_line(line, sep)
        obj: ToonRow = {}
        for i, key in enumerate(schema):
            obj[key] = parts[i] if i < len(parts) else ""
        rows.append(obj)

    return rows


def is_valid_toon(toon: str, expected_header: str) -> bool:
    """
    Guardrail: check that a TOON string starts with EXACT expected header.
    """
    if not toon:
        return False
    first_line = toon.replace("\r", "").split("\n")[0].strip()
    return first_line == expected_header.strip()


def estimate_savings(rows: List[ToonRow], schema: ToonSchema) -> Dict[str, Any]:
    """
    Rough token-savings estimator (characters as proxy).
    """
    import json
    json_str = json.dumps(rows)
    toon_str = encode_toon(rows, schema)
    json_chars = len(json_str)
    toon_chars = len(toon_str)
    saved_pct = 0
    if json_chars > 0:
        saved_pct = max(0, round((1 - toon_chars / json_chars) * 100))
    return {
        "jsonChars": json_chars,
        "toonChars": toon_chars,
        "savedPct": saved_pct,
    }

File: src/lib/toon.ts
/**
 * TOON: Tabular-Object Optimized Notation
 * Lightweight, column-ordered rows with a tiny header.
 *
 * Format:
 *   #TOON v=1 sep=; cols=item,qty,length,width,room
 *   Upper cabinet;4;36;12;Breakroom 204
 *   Base cabinet;2;48;24;Breakroom 204
 *
 * - First line starts with "#TOON" and declares version, separator, and column order.
 * - Fields are joined with the chosen separator (default ";").
 * - If a field contains separator/newline/#/quote, it is wrapped with double quotes,
 *   and quotes inside are doubled (" -> "").
 */

export type ToonRow = Record<string, string | number | boolean | null | undefined>;
export type ToonSchema = string[]; // ordered list of fields

const DEFAULT_SEP = ";";

/** Escape a single field for TOON */
function escapeField(raw: unknown, sep = DEFAULT_SEP): string {
  const s = raw === null || raw === undefined ? "" : String(raw);
  const needsQuote =
    s.includes(sep) ||
    s.includes("\n") ||
    s.includes("\r") ||
    s.includes('"') ||
    s.startsWith("#");

  if (!needsQuote) return s;
  return `"${s.replace(/"/g, '""')}"`;
}

/** Unescape a single TOON field back to plain string */
function unescapeField(s: string): string {
  if (s.startsWith('"') && s.endsWith('"')) {
    return s.slice(1, -1).replace(/""/g, '"');
  }
  return s;
}

/**
 * JSON[] -> TOON string
 *
 * @param rows   Array of row objects
 * @param schema Ordered list of keys (columns)
 * @param sep    Column separator (default ";")
 */
export function encodeToon(rows: ToonRow[], schema: ToonSchema, sep = DEFAULT_SEP): string {
  const header = `#TOON v=1 sep=${sep} cols=${schema.join(",")}`;
  const body = rows
    .map((r) => schema.map((k) => escapeField(r[k], sep)).join(sep))
    .join("\n");
  return `${header}\n${body}`;
}

/**
 * TOON string -> JSON[]
 *
 * Throws if the header is missing/invalid.
 */
export function decodeToon(toon: string): ToonRow[] {
  const lines = toon.replace(/\r/g, "").split("\n");
  if (lines.length === 0 || !lines[0].startsWith("#TOON")) {
    throw new Error("Invalid TOON header");
  }

  const header = lines[0];
  const colsMatch = header.match(/cols=([^\s]+)/);
  const sepMatch = header.match(/sep=([^\s]+)/);
  if (!colsMatch || !sepMatch) {
    throw new Error("Malformed TOON header");
  }

  const schema = colsMatch[1].split(",");
  const sep = sepMatch[1];

  // Split a line into fields, respecting quotes and escaped quotes
  function splitLine(line: string): string[] {
    const out: string[] = [];
    let cur = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
      const c = line[i];

      if (inQuotes) {
        if (c === '"') {
          if (line[i + 1] === '"') {
            cur += '"';
            i++; // skip escaped quote
          } else {
            inQuotes = false;
          }
        } else {
          cur += c;
        }
        continue;
      }

      if (c === '"') {
        inQuotes = true;
        continue;
      }

      // NOTE: sep currently assumed 1 char (e.g. ";")
      if (sep.length === 1 && c === sep) {
        out.push(cur);
        cur = "";
      } else {
        cur += c;
      }
    }

    out.push(cur);
    return out.map(unescapeField);
  }

  const rows: ToonRow[] = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const parts = splitLine(line);
    const obj: ToonRow = {};
    for (let j = 0; j < schema.length; j++) {
      obj[schema[j]] = parts[j] ?? "";
    }
    rows.push(obj);
  }

  return rows;
}

/**
 * Helper guardrail:
 * Check that a returned TOON string starts with the EXACT expected header.
 *
 * Use this right after an LLM call before trusting / decoding the output.
 */
export function isValidToon(toon: string, expectedHeader: string): boolean {
  if (!toon) return false;
  const firstLine = toon.replace(/\r/g, "").split("\n")[0]!.trim();
  return firstLine === expectedHeader.trim();
}

/**
 * Rough token-savings estimator (chars as proxy).
 * Not perfect, but good enough to see the win.
 */
export function estimateSavings(
  jsonRows: ToonRow[],
  schema: ToonSchema
): { jsonChars: number; toonChars: number; savedPct: number } {
  const json = JSON.stringify(jsonRows);
  const toon = encodeToon(jsonRows, schema);
  const jsonChars = json.length;
  const toonChars = toon.length;
  const savedPct = jsonChars
    ? Math.max(0, Math.round((1 - toonChars / jsonChars) * 100))
    : 0;

  return { jsonChars, toonChars, savedPct };
}
